In [59]:
import numpy as np
import pandas as pd
from warnings import filterwarnings
filterwarnings("ignore")
import seaborn as sns
from scipy import stats

Используя данные о хоккеистах (скачать можно здесь), проверьте, является ли среди хоккеистов из Финляндии, Норвегии и Дании значимым отличие: а) роста, б) BMI.

In [47]:
df = pd.read_csv('hockey_players.csv', encoding='ISO-8859-1')
df.sample(5)

year country  no              name position side  height  weight  \
4778  2013     CAN   2       hamhuis dan        D    L     186    92.0   
3726  2010     ITA  28    de toni manuel        F    L     181    80.0   
129   2001     FIN  15  laaksonen antti         F    L     183    82.0   
5056  2013     SWE  81       falth elias        D    R     176    79.0   
4071  2011     FIN  47       lahti janne        F    L     189    89.0   

           birth               club        age  cohort        bmi  
4778  1982-12-13  vancouver canucks  30.381930    1982  26.592670  
3726  1979-01-10     alleghe hockey  31.304586    1979  24.419279  
129   1973-10-03    minnesota wild   27.575633    1973  24.485652  
5056  1981-03-30     hv71 jonkoping  32.087611    1981  25.503616  
4071  1982-07-20   jokerit helsinki  28.780287    1982  24.915316

In [48]:
df = df.sort_values(by='age', ascending=True,  ignore_index=True).drop_duplicates(subset='name', keep='last')

In [49]:
df.describe()

year           no       height       weight          age  \
count  3278.000000  3278.000000  3278.000000  3278.000000  3278.000000   
mean   2008.335570    29.231239   183.893228    87.945546    27.859656   
std       5.093477    23.271197     5.480919     7.045355     4.403044   
min    2001.000000     1.000000   165.000000    65.000000    17.686516   
25%    2004.000000    12.000000   180.000000    83.000000    24.681725   
50%    2007.000000    23.000000   184.000000    88.000000    27.626283   
75%    2013.000000    38.000000   188.000000    93.000000    30.839151   
max    2016.000000    97.000000   205.000000   117.000000    43.206023   

            cohort          bmi  
count  3278.000000  3278.000000  
mean   1980.359365    25.989036  
std       6.808705     1.481745  
min    1963.000000    19.460881  
25%    1975.000000    25.057360  
50%    1980.000000    25.978680  
75%    1985.000000    26.880935  
max    1998.000000    30.930438

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3278 entries, 1 to 6291
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      3278 non-null   int64  
 1   country   3278 non-null   object 
 2   no        3278 non-null   int64  
 3   name      3278 non-null   object 
 4   position  3278 non-null   object 
 5   side      3277 non-null   object 
 6   height    3278 non-null   int64  
 7   weight    3278 non-null   float64
 8   birth     3278 non-null   object 
 9   club      3278 non-null   object 
 10  age       3278 non-null   float64
 11  cohort    3278 non-null   int64  
 12  bmi       3278 non-null   float64
dtypes: float64(3), int64(4), object(6)
memory usage: 358.5+ KB


In [52]:
df.country.value_counts()

USA    297
CAN    269
SWE    238
RUS    219
CZE    204
FIN    200
SVK    189
GER    179
SUI    163
BLR    155
LAT    138
AUT    136
DEN    132
KAZ    124
UKR    114
SLO    113
ITA    102
NOR     98
FRA     73
JPN     69
HUN     43
POL     23
Name: country, dtype: int64

In [57]:
fin_h = df.loc[df.country == 'FIN'].height
nor_h = df.loc[df.country == 'NOR'].height
den_h = df.loc[df.country == 'DEN'].height

fin_bmi = df.loc[df.country == 'FIN'].bmi
nor_bmi = df.loc[df.country == 'NOR'].bmi
den_bmi = df.loc[df.country == 'DEN'].bmi

In [63]:
k = 3
n = 3278 * 3

k1 = k - 1
k2 = n - k

In [66]:
alpha = 0.05

t = stats.f.ppf(1 - alpha, k1, k2)
t

2.9966453276970184

Итак, критическая область имеет вид:
$$\Omega_\alpha = \left( 2.997, +\infty \right)$$

In [67]:
stats.f_oneway(fin_h, nor_h, den_h)

F_onewayResult(statistic=3.734875379545475, pvalue=0.024659740563714175)

In [68]:
stats.f_oneway(fin_bmi, nor_bmi, den_bmi)

F_onewayResult(statistic=5.532919076049783, pvalue=0.004243187723739065)

Значения статистик попали в критическую область, поэтому заключаем, что отличие роста и bmi действительно являются статистически значимыми.